### ISSUE TITLE
Flat rate discounts do not consider quantity

### ISSUE THREAD

**Issue by brashrebel** [easydigitaldownloads/easy-digital-downloads#5304]
Reported by a customer and just confirmed in my test. A flat rate discount only deducts its value from the sum of the single items in the cart. But any additional quantity are not considered in the calculation.

Example:
- Discount code good for $10
- Item in the cart costs $1
- Item quantity is 5
- Amount discounted = $1

**Comment by sunnyratilal** [comment]
Off the top of my head, I have a hunch that this can be simply fixed by multiplying the discounted amount to the value of the quantity stored in the cart contents but we may need to do additional checks for max uses.

E.g. If the discount code is only specific for a product and has a max usage of 3 - does this usage count update based on quantity or just as one use albeit different quantities? So based on @brashrebel's example: would the usage of the discount code would change to 5 or stay as 1?

**Comment by chriscct7** [comment]
I would think based on quantity. Which also means if the usage limit is 3
and you've used 1 already, if you add 3 of an item to cart, the discount
should only apply 2 times

**Comment by brashrebel** [comment]
I think quantity should only apply if it is a download specific discount and not a cart discount. If it is a discount on the entire order than the limit number should be the number of transactions. If it is restricted to a specific download, than I agree with Chris that quantity makes more sense.

**Comment by pippinsplugins** [comment]
I don't think we currently have enough settings within the discount screen to account for this. We'll need to introduce discount types (or similar) before we can handle these various use cases.

Currently, a discount can be limited to specific products but doing that doesn't mean the behavior of the discount changes in any way.

We can't just assume that a flat rate discount should be multiplied by item quantity since we really don't have a way of determining whether it's a product or a cart discount. Just because a discount is limited to one or more products does not necessarily make it a product discount.

I've pushed this back to 2.8 and propose we do the following:

1. Add a field for setting the discount type, `product discount` and `cart discount`
2. When type is `product discount`, flat rate discounts get multiplied by the quantity
3. When type is `cart discount`, behavior stays as it is now

**Comment by chriscct7** [comment]
It's a product discount if the discount screen is limited to specific
products. It's already distinguishable

**Comment by chriscct7** [comment]
> there's nothing to distinguish between a discount restricted to one product or five products.

We store this information in the UI for edit discounts?

**Comment by pippinsplugins** [comment]
We store the product limitations, yes.

I'm trying to say that the UI for product limitations does not inherently make a distinction between discount types.

A discount that works on specific products !== to a product discount type.

**Comment by chriscct7** [comment]
We can add a more specific sentence to the UI. "If a product is specified in the requirements section above than the limit is calculated by x. If there is no product specified in the requirements section above then the limit is calculated by y". We can even use Javascript to detect the requirements section having a value or not and replace the description text to which use case is being used. It's the far better, simple solution that keeps EDD focused, solves the problem, and allows other things to move forward.

We already have an extension for different discount types (Discounts Pro), that does all sorts of discount types. There's a reason it requires thousands of lines of code. The more features you add into discounts, the more complicated things get (exponentially). Not just in the calculation of the discount or the setting up of the discount, but also in terms of integration with fees, taxes, shipping, etc. EDD core should just stick to the main use cases of discounts, the ones 80% of people will use. Its not even just core cases, when EDD adds more features to discounts it affects other extensions, including for example the VAT extension, and then we'll get people complaining things aren't VAT compliant. The list goes on. 

The discounts in EDD are already broken in many cases because of how many more features they've gotten over the last couple of years. For example, there's already **17 currently open outstanding bugs** with discounts, and there's definitely an issue somewhere inside of allow product quantity to be edited turned off but allowing product quantities, as it miscalculates taxes. As soon as I figure out what causes it, that'll be 18 tickets.

No one is asking EDD for more discounts, rather users would prefer discounts stay simpler, but work 100% of the time.

**Comment by pippinsplugins** [comment]
Just adding a sentence is not enough.

Changing the behavior here could potentially have a drastic effect on existing sites. For example, someone might have a $5 flat discount set up that gives $5 off an order, regardless of the quantity ordered. If we change it to apply X times based on the quantity, that discount could suddenly be $50 discount or more.

To do this, we have to add a new option to the discount screen.

**Comment by nikovermeer** [comment]
Until somewhere after February 2016 and before November 2016 the discount code flat amount actually DID apply to fees as well. I'm 100% certain of this. Since November I'm using a different bug to be able to book discounted tickets. 

Same applies to https://github.com/easydigitaldownloads/easy-digital-downloads/issues/5301 I believe.

Would very much appreciate it if discounts are applied to cart by default again, or options are provided.

**Comment by pippinsplugins** [comment]
@nikovermeer discount codes have never applied to fees.

**Comment by pippinsplugins** [comment]
I've had another report of this affecting a customer: https://secure.helpscout.net/conversation/331701347/54852/?folderId=180505

Their item costs $57, their discount set to $1781, and a purchase containing 50 copies of the $57 item only gets discounted by $57.

**Comment by cklosowski** [comment]
Related #5879

**Comment by dgoldak** [comment]
Another report of this affecting a customer: https://secure.helpscout.net/conversation/814237237/103943?folderId=180499
